README.rst
==========
musicburst
==========

This is a simple tool for extracting data from EAF files for analysis of the
``MusicBurst`` and ``Source`` tiers, and writing the output to a CSV file, with
one row per input file.

Installation
============

Clone the repository, then use ``pip`` to install the ``musicburst`` tool::

  $ git clone https:://github.com/gedankenexperimenter/musicburst

  $ pip install musicburst

This should result in a ``musicburst`` command line program becoming available
in your path. In a directory with EAF files containing the target data, then
run::

  $ musicburst *.eaf

The output data will be written to the file ``musicburst-counts.csv``,
containing columns for:

- file name
- total length (until the end of the last annotated segment, which is not
  necessary the total length of the source WAV file)
- number of segments in the ``MusicBurst`` tier
- total length of those ``MusicBurst`` segments
- number of segments of the ``Source`` tier with the annotation ``1``
- total length of those ``Source`` tier segments

Notes
=====

- All time values are in milliseconds.
- Total length of the recording is not available from the EAF file alone. If you
  need the true total length, you'll have to get that from the WAV file.


CHANGELOG.rst
=========
Changelog
=========

Version 1.0.0
=============

- Testing is now working
- Cleaned up most pylint warnings


Version 0.1.0
=============

- All planned features working


tests/conftest.py
"""
    Dummy conftest.py for musicburst.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    - https://docs.pytest.org/en/stable/fixture.html
    - https://docs.pytest.org/en/stable/writing_plugins.html
"""

# import pytest


tests/test_main.py
"""
Tests for main() function
"""
import pytest

from musicburst.main import main

__author__ = "Michael Richters"
__copyright__ = "Michael Richters"
__license__ = "MIT"

HEADERS = [
    'filename', 'total time',
    'music segments', 'music time',
    'singing segments', 'singing time',
]

def test_main(capsys):
    """CLI Tests"""
    # capsys is a pytest fixture that allows asserts agains stdout/stderr
    # https://docs.pytest.org/en/stable/capture.html
    main(['-o', '-', 'tests/test.eaf'])
    captured = capsys.readouterr()
    csv_output = captured.out.split('\n')
    for line in csv_output:
        print(line)
    assert csv_output[0] == ','.join(HEADERS)
    assert csv_output[1] == "test,7320009,48,638273,18,128369"


tests/test_skeleton.py
import pytest

from musicburst.skeleton import fib, main

__author__ = "Michael Richters"
__copyright__ = "Michael Richters"
__license__ = "MIT"


def test_fib():
    """API Tests"""
    assert fib(1) == 1
    assert fib(2) == 1
    assert fib(7) == 13
    with pytest.raises(AssertionError):
        fib(-10)


def test_main(capsys):
    """CLI Tests"""
    # capsys is a pytest fixture that allows asserts agains stdout/stderr
    # https://docs.pytest.org/en/stable/capture.html
    main(["7"])
    captured = capsys.readouterr()
    assert "The 7-th Fibonacci number is 13" in captured.out


src/musicburst/skeleton.py
"""
This is a skeleton file that can serve as a starting point for a Python
console script. To run this script uncomment the following lines in the
``[options.entry_points]`` section in ``setup.cfg``::

    console_scripts =
         fibonacci = musicburst.skeleton:run

Then run ``pip install .`` (or ``pip install -e .`` for editable mode)
which will install the command ``fibonacci`` inside your current environment.

Besides console scripts, the header (i.e. until ``_logger``...) of this file can
also be used as template for Python modules.

Note:
    This skeleton file can be safely removed if not needed!

References:
    - https://setuptools.readthedocs.io/en/latest/userguide/entry_point.html
    - https://pip.pypa.io/en/stable/reference/pip_install
"""

import argparse
import logging
import sys

from musicburst import __version__

__author__ = "Michael Richters"
__copyright__ = "Michael Richters"
__license__ = "MIT"

_logger = logging.getLogger(__name__)


# ---- Python API ----
# The functions defined in this section can be imported by users in their
# Python scripts/interactive interpreter, e.g. via
# `from musicburst.skeleton import fib`,
# when using this Python module as a library.


def fib(n):
    """Fibonacci example function

    Args:
      n (int): integer

    Returns:
      int: n-th Fibonacci number
    """
    assert n > 0
    a, b = 1, 1
    for i in range(n - 1):
        a, b = b, a + b
    return a


# ---- CLI ----
# The functions defined in this section are wrappers around the main Python
# API allowing them to be called directly from the terminal as a CLI
# executable/script.


def parse_args(args):
    """Parse command line parameters

    Args:
      args (List[str]): command line parameters as list of strings
          (for example  ``["--help"]``).

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(description="Just a Fibonacci demonstration")
    parser.add_argument(
        "--version",
        action="version",
        version="musicburst {ver}".format(ver=__version__),
    )
    parser.add_argument(dest="n", help="n-th Fibonacci number", type=int, metavar="INT")
    parser.add_argument(
        "-v",
        "--verbose",
        dest="loglevel",
        help="set loglevel to INFO",
        action="store_const",
        const=logging.INFO,
    )
    parser.add_argument(
        "-vv",
        "--very-verbose",
        dest="loglevel",
        help="set loglevel to DEBUG",
        action="store_const",
        const=logging.DEBUG,
    )
    return parser.parse_args(args)


def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(
        level=loglevel, stream=sys.stdout, format=logformat, datefmt="%Y-%m-%d %H:%M:%S"
    )


def main(args):
    """Wrapper allowing :func:`fib` to be called with string arguments in a CLI fashion

    Instead of returning the value from :func:`fib`, it prints the result to the
    ``stdout`` in a nicely formated message.

    Args:
      args (List[str]): command line parameters as list of strings
          (for example  ``["--verbose", "42"]``).
    """
    args = parse_args(args)
    setup_logging(args.loglevel)
    _logger.debug("Starting crazy calculations...")
    print("The {}-th Fibonacci number is {}".format(args.n, fib(args.n)))
    _logger.info("Script ends here")


def run():
    """Calls :func:`main` passing the CLI arguments extracted from :obj:`sys.argv`

    This function can be used as entry point to create console scripts with setuptools.
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    # ^  This is a guard statement that will prevent the following code from
    #    being executed in the case someone imports this file instead of
    #    executing it as a script.
    #    https://docs.python.org/3/library/__main__.html

    # After installing your project with pip, users can also run your Python
    # modules as scripts via the ``-m`` flag, as defined in PEP 338::
    #
    #     python -m musicburst.skeleton 42
    #
    run()


src/musicburst/main.py
# -*- coding: utf-8 -*-
"""
This is a script that parses EAF files and counts segments of the 'MusicBurst'
tier, and segments of the 'Source' tier annotated with the value '1'. It takes a
set of EAF files as input, and produces a CSV file as output, with columns for
file name, total file length (from beginning of the recording to the end of the
last annotated segment; the original WAV file may be longer), number of segments
in the 'MusicBurst' tier, total length of time of 'MusicBurst' tier segments
summed, number of segments in the 'Source' tier annotated with the value '1',
and the total length of those segments.
"""

import argparse
import csv
import logging
import os
import sys

from collections import defaultdict
from contextlib import redirect_stdout

# pympi-ling is required for parsing of EAF files
import pympi

from musicburst import __version__

__author__ = "Michael Richters"
__copyright__ = "Michael Richters"
__license__ = "MIT"

# ==============================================================================
# Constants
MUSIC_TIER_NAME = 'MusicBurst'
SOURCE_TIER_NAME = 'Source'

# ==============================================================================
class OutputRecord:
    """Represents a row of the data table to be written to the output file"""
    data_labels = ['total time',
                   'music segments', 'music time',
                   'singing segments', 'singing time']
    header = ['filename']
    header.extend(data_labels)

    def __init__(self, filename):
        self.filename = os.path.basename(filename).replace('.eaf', '')
        self.data = defaultdict(int)

    def fmt(self):
        """Format the output record, returning a list of values, with zeros replaced by
        empty strings."""
        values = [self.filename]
        def _blank_zero(entry):
            value = self.data[entry]
            return '' if value == 0 else value
        data_values = map(_blank_zero, self.data_labels)
        values.extend(data_values)
        return values

# ------------------------------------------------------------------------------
class Error(Exception):
    """Base class for errors in this module"""

class InputError(Error):
    """Exception raised for errors that occur when reading input

    Attributs:
        message (string): an description of the error condition
    """
    def __init__(self, message):
        super().__init__()
        self.message = message

# ==============================================================================
def parse_args(args):
    """Parse command line parameters

    Args:
      args (List[str]): command line parameters as list of strings
          (for example  ``["--help"]``).

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(
        description="Generate summary of MusicBurst  tier data in EAF file(s)"
    )
    parser.add_argument(
        '--version',
        action  = 'version',
        version = "musicburst {ver}".format(ver=__version__),
    )
    parser.add_argument(
        '-v', '--verbose',
        dest    = 'loglevel',
        help    = "set loglevel to INFO",
        action  = 'store_const',
        const   = logging.INFO,
    )
    parser.add_argument(
        '-vv', '--very-verbose',
        dest    = 'loglevel',
        help    = "set loglevel to DEBUG",
        action  = 'store_const',
        const   = logging.DEBUG,
    )
    parser.add_argument(
        '-o', '--output',
        metavar = '<csv_file>',
        type    = argparse.FileType('w'),
        default = 'musicburst-counts.csv',
        help    = "Write output to <csv_file> (default: '%(default)s')",
    )
    parser.add_argument(
        '-d', '--delimiter',
        choices = ['tab', 'comma', 'ascii'],
        default = 'comma',
        help    = "Use <delimiter> as CSV output field separator (default: '%(default)s')",
    )
    parser.add_argument(
        'eaf_files',
        metavar = '<eaf_file>',
        nargs   = '+',
        help    = "The name(s) of the EAF file(s) to process",
    )
    return parser.parse_args(args)

# ==============================================================================
def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "%(levelname)s: %(message)s"
    logging.basicConfig(
        level   = loglevel,
        stream  = sys.stdout,
        format  = logformat,
        datefmt = "%Y-%m-%d %H:%M:%S",
    )

# ==============================================================================
def setup_output(filename, delimiter):
    """Setup CSV output and write header row

    Args:
      :obj:`argparse.Namespace`: command line parameters namespace

    Returns:
      :obj:`csv.writer`: CSV output object
    """

    # Create the CSV writer object as specified by args
    output = csv.writer(filename,
                        delimiter      = delimiter,
                        quoting        = csv.QUOTE_MINIMAL,
                        lineterminator = '\n')
    # Write the header row
    logging.debug("Writing output header")
    output.writerow(OutputRecord.header)
    return output

# ==============================================================================
def collect_music_data(eaf_file):
    """Collect data from an EAF file

    Args:
      eaf_file (string): filename of an EAF file to analyze

    Returns:
      :obj:`OutputRecord`: record containing the output data
    """
    logging.info("Processing %s", eaf_file)

    # pympi prints warnings about EAF versions to STDOUT instead of STDERR. This
    # gets in the way of CSV output to STDOUT.
    with redirect_stdout(sys.stderr):
        eaf = pympi.Elan.Eaf(eaf_file)

    tier_names = eaf.get_tier_names()
    logging.debug("All tiers: %s", format(list(tier_names)))

    output_record = OutputRecord(eaf_file)

    if MUSIC_TIER_NAME not in tier_names:
        raise InputError("Missing {} tier in file {}"
                         .format(MUSIC_TIER_NAME, eaf_file))

    for record in eaf.get_annotation_data_for_tier(MUSIC_TIER_NAME):
        (start, end, value) = record[:3]
        logging.debug("%s segment: %s",
                      MUSIC_TIER_NAME, format((start, end, value)))
        output_record.data['music segments'] += 1
        output_record.data['music time'] += (end - start)

    if SOURCE_TIER_NAME not in tier_names:
        logging.warning("Missing source tier in file %s", eaf_file)
    else:
        for record in eaf.get_annotation_data_for_tier('Source'):
            (start, end, value) = record[:3]
            logging.debug("%s segment: %s",
                          SOURCE_TIER_NAME, format((start, end, value)))
            if value != '1':
                continue
            output_record.data['singing segments'] += 1
            output_record.data['singing time'] += (end - start)

    (start, end) = eaf.get_full_time_interval()
    output_record.data['total time'] = (end - start)

    return output_record

# ==============================================================================
def main(args):
    """Command-line interface function for the script to parse EAF files for
    segments in the 'MusicBurst' tier and 'Source' tier. Calls
    ``collect_music_data()`` for each EAF file specified on the command line,
    and writes the data to a specified CSV output file.

    Args:
      args (List[str]): command line parameters as list of strings
          (for example  ``["--verbose", "42"]``).
    """
    args = parse_args(args)
    setup_logging(args.loglevel)

    # First, set the output delimiter character from args
    output_delimiter = '\t'
    if args.delimiter == 'comma':
        output_delimiter = ','
    elif args.delimiter == 'ascii':
        output_delimiter = '\x1f'

    output = setup_output(args.output, output_delimiter)

    for eaf_file in args.eaf_files:
        try:
            output_record = collect_music_data(eaf_file)
        except InputError as err:
            logging.warning(err.message)
            continue

        output.writerow(output_record.fmt())

    if args.output != sys.stdout:
        args.output.close()


def run():
    """Calls :func:`main` passing the CLI arguments extracted from :obj:`sys.argv`

    This function can be used as entry point to create console scripts with setuptools.
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    # ^  This is a guard statement that will prevent the following code from
    #    being executed in the case someone imports this file instead of
    #    executing it as a script.
    #    https://docs.python.org/3/library/__main__.html

    # After installing your project with pip, users can also run your Python
    # modules as scripts via the ``-m`` flag, as defined in PEP 338::
    #
    #     python -m musicburst.skeleton 42
    #
    run()
